File: typescript/agent.ts
/**
 * Procurement service agent — TypeScript example.
 *
 * Validates purchase requests and resumes. Workflow then pauses
 * for manager approval.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   npx tsx agent.ts
 */

import { AxmeClient } from "@axme/axme";

const AGENT_ADDRESS = "procurement-service-demo";

async function handleIntent(client: AxmeClient, intentId: string) {
  const intentData = await client.getIntent(intentId);
  const intent = intentData.intent ?? intentData;
  let payload = intent.payload ?? {};
  if (payload.parent_payload) {
    payload = payload.parent_payload;
  }

  const requestId = payload.request_id ?? "unknown";
  const amount = payload.amount ?? 0;
  const dept = payload.department ?? "unknown";

  console.log(`  Processing purchase ${requestId}: $${amount} for ${dept}...`);
  await new Promise((r) => setTimeout(r, 1000));
  console.log(`  Validating budget availability...`);
  await new Promise((r) => setTimeout(r, 1000));

  const result = {
    action: "complete",
    request_id: requestId,
    budget_available: true,
    validated_at: new Date().toISOString(),
  };

  await client.resumeIntent(intentId, result, { ownerAgent: "procurement-service-demo" });
  console.log(`  Purchase ${requestId} validated. Waiting for manager approval.`);
  console.log(`  To approve: axme tasks approve <intent_id>`);
}

async function main() {
  const apiKey = process.env.AXME_API_KEY;
  if (!apiKey) {
    console.error("Error: AXME_API_KEY not set.");
    process.exit(1);
  }

  const client = new AxmeClient({ apiKey });

  console.log(`Agent listening on ${AGENT_ADDRESS}...`);
  console.log("Waiting for intents (Ctrl+C to stop)\n");

  for await (const delivery of client.listen(AGENT_ADDRESS)) {
    const intentId = delivery.intent_id;
    const status = delivery.status;
    if (intentId && ["DELIVERED", "CREATED", "IN_PROGRESS"].includes(status)) {
      console.log(`[${status}] Intent received: ${intentId}`);
      try {
        await handleIntent(client, intentId);
      } catch (e) {
        console.error(`  Error: ${e}`);
      }
    }
  }
}

main().catch(console.error);

File: java/Agent.java
/*
 * Procurement service agent — Java example.
 *
 * Fetches an intent by ID, validates a purchase request, and resumes.
 * Workflow then pauses for manager approval.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   javac -cp axme-sdk.jar Agent.java
 *   java -cp .:axme-sdk.jar Agent <intent_id>
 */

import dev.axme.sdk.AxmeClient;
import dev.axme.sdk.AxmeClientConfig;
import dev.axme.sdk.RequestOptions;
import java.time.Instant;
import java.util.Map;

public class Agent {
    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java Agent <intent_id>");
            System.exit(1);
        }

        String apiKey = System.getenv("AXME_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            System.err.println("Error: AXME_API_KEY not set.");
            System.exit(1);
        }

        String intentId = args[0];
        var client = new AxmeClient(AxmeClientConfig.forCloud(apiKey));

        System.out.println("Processing intent: " + intentId);

        var intentData = client.getIntent(intentId, new RequestOptions());
        @SuppressWarnings("unchecked")
        var intent = (Map<String, Object>) intentData.getOrDefault("intent", intentData);
        @SuppressWarnings("unchecked")
        var payload = (Map<String, Object>) intent.getOrDefault("payload", Map.of());
        if (payload.containsKey("parent_payload")) {
            @SuppressWarnings("unchecked")
            var pp = (Map<String, Object>) payload.get("parent_payload");
            payload = pp;
        }

        String requestId = (String) payload.getOrDefault("request_id", "unknown");
        double amount = payload.containsKey("amount") ? ((Number) payload.get("amount")).doubleValue() : 0;
        String dept = (String) payload.getOrDefault("department", "unknown");

        System.out.printf("  Processing purchase %s: $%.0f for %s...%n", requestId, amount, dept);
        Thread.sleep(1000);
        System.out.println("  Validating budget availability...");
        Thread.sleep(1000);

        var result = Map.<String, Object>of(
            "action", "complete",
            "request_id", requestId,
            "budget_available", true,
            "validated_at", Instant.now().toString()
        );

        client.resumeIntent(intentId, result, new RequestOptions());
        System.out.println("  Purchase " + requestId + " validated. Waiting for manager approval.");
        System.out.println("  To approve: axme tasks approve <intent_id>");
    }
}
